### PR TITLE
feature(whl): add demo code for lstm and gtrxl

### PR DESCRIPTION
**Pull request by kxzxvbk**

### CODE REVIEW

File: chapter5_time/gtrxl.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+import torch
+import torch.nn as nn
+from ding.torch_utils.network.nn_module import fc_block
+from ding.torch_utils.network.gtrxl import PositionalEmbedding, GatedTransformerXLLayer, Memory
+
+
+class GTrXL(nn.Module):
+    """
+    **Overview:**
+        Pytorch implementation for GTrXL.
```
Comment (PaParaZz1) [inline chapter5_time/gtrxl.py:19]: PyTorch

File: chapter5_time/gtrxl.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        # Return memory is needed.
+        if return_mem:
+            output = {"logit": out, "memory": memory}
+        else:
+            output = {"logit": out}
+        return output
+
+
+if __name__ == '__main__':
+    # Generate data for testing.
```
Comment (PaParaZz1) [inline chapter5_time/gtrxl.py:171]: move to `test_gtrxl` function

File: chapter5_time/gtrxl.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        # Get full sequence length: memory length + current length
+        prev_seq = self.memory_len
+        full_seq = cur_seq + prev_seq
+        # If the attention mask for current sequence length is already created, reuse the mask stored in self.att_mask.
+        if cur_seq in self.att_mask.keys():
+            attn_mask = self.att_mask[cur_seq]
+        # Otherwise, create a new attention mask and store it into self.att_mask.
+        else:
+            attn_mask = (
+                torch.triu(
```
Comment (PaParaZz1) [inline chapter5_time/gtrxl.py:126]: why create mask like this, add some comments

File: chapter5_time/gtrxl.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            hidden_state.append(out.clone())
+        out = self.dropout(out)
+        # Update the GTrXL memory.
+        self.memory.update(hidden_state)
+        # If the first dimension of output is required to be batch_size, then reshape x from  [sequence_length, batch_size, input_dim] to [batch_size ,sequence_length ,input_dim].
+        if batch_first:
+            out = torch.transpose(out, 1, 0)
+        # Return memory is needed.
+        if return_mem:
+            output = {"logit": out, "memory": memory}
```
Comment (PaParaZz1) [inline chapter5_time/gtrxl.py:164]: return treetensor

File: chapter5_time/lstm.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            self.wh.append(nn.Parameter(torch.zeros(hidden_size, hidden_size * 4)))
+        self.bias = nn.Parameter(torch.zeros(num_layers, hidden_size * 4))
+        # Initialize the Dropout Layer.
+        self.use_dropout = dropout > 0.
+        if self.use_dropout:
+            self.dropout = nn.Dropout(dropout)
+        self._init()
+
+    # Dealing with different types of input and return preprocessed prev_state.
+    def _before_forward(self, inputs: torch.Tensor, prev_state: Union[None, List[Dict]]) -> torch.Tensor:
```
Comment (PaParaZz1) [inline chapter5_time/lstm.py:53]: remove before and after forward function, keep the similar input/output types with `nn.LSTM`

File: chapter5_time/lstm.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    **Overview:**
+        Implementation of LSTM cell with layer norm.
+    """
+
+    def __init__(
+            self,
+            input_size: int,
+            hidden_size: int,
+            num_layers: int,
+            norm_type: Optional[str] = None,
```
Comment (PaParaZz1) [inline chapter5_time/lstm.py:27]: defaults to `LN`

File: chapter5_time/lstm.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            state = []
+            for prev in prev_state:
+                state.append([v for v in prev.values()])
+            state = list(zip(*state))
+            prev_state = [torch.cat(t, dim=1) for t in state]
+
+        return prev_state
+
+    def _init(self):
+        # Initialize parameters.
```
Comment (PaParaZz1) [inline chapter5_time/lstm.py:79]: add more explanation about this initialization

File: chapter5_time/gtrxl.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+- Pytorch implementation for GTrXL.
+- An example to test GTrXL.
+"""
+from typing import Optional, Dict
+import warnings
+import numpy as np
+import torch
+import torch.nn as nn
+from ding.torch_utils.network.nn_module import fc_block
+from ding.torch_utils.network.gtrxl import PositionalEmbedding, GatedTransformerXLLayer, Memory
```
Comment (PaParaZz1) [inline chapter5_time/gtrxl.py:13]: move `GatedTransformerXLLayer` to this file

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]
